mds/src/karatsuba_convolution.rs

```diff
@@ -0,0 +1,362 @@
+
+        // right = lhs(x)rhs(x) mod x^{N/2} - 1
+        inner_conv(lhs_pos, rhs_pos, right);
+
+        for i in 0..HALF_N {
+            left[i] += right[i]; // w_0 + w_1
+            left[i] >>= 1; // (w_0 + w_1)/2
+            right[i] -= left[i]; // (w_0 - w_1)/2
+        }
+    }
+
+    /// Compute output(x) = lhs(x)rhs(x) mod x^N + 1.
+    ///
+    /// NB: This function should not be called by clients.
+    #[inline(always)]
+    fn signed_conv_n<
+        const N: usize,
+        const HALF_N: usize,
+        SC: Fn([T; HALF_N], [U; HALF_N], &mut [V]),
+    >(
+        lhs: [T; N],
+        rhs: [U; N],
+        output: &mut [V],
+        inner_signed_conv: SC,
+    ) {
+        debug_assert_eq!(2 * HALF_N, N);
+        // NB: The compiler is smart enough not to initialise these arrays.
+        let mut lhs_even = [T::default(); HALF_N];
+        let mut lhs_odd = [T::default(); HALF_N];
+        let mut lhs_sum = [T::default(); HALF_N];
+        let mut rhs_even = [U::default(); HALF_N];
+        let mut rhs_odd = [U::default(); HALF_N];
+        let mut rhs_sum = [U::default(); HALF_N];
+
+        for i in 0..HALF_N {
+            let s = lhs[2 * i];
+            let t = lhs[2 * i + 1];
+            lhs_even[i] = s;
+            lhs_odd[i] = t;
+            lhs_sum[i] = s + t;
+
+            let s = rhs[2 * i];
+            let t = rhs[2 * i + 1];
+            rhs_even[i] = s;
+            rhs_odd[i] = t;
+            rhs_sum[i] = s + t;
+        }
+
+        let mut even_s_conv = [V::default(); HALF_N];
+        let (left, right) = output.split_at_mut(HALF_N);
+
+        inner_signed_conv(lhs_even, rhs_even, &mut even_s_conv);
+        inner_signed_conv(lhs_odd, rhs_odd, left);
+        inner_signed_conv(lhs_sum, rhs_sum, right);
+
+        // Adjust so that the correct values are in right and
+        // even_s_conv respectively:
+        right[0] -= even_s_conv[0] + left[0];
+        even_s_conv[0] -= left[HALF_N - 1];
+
+        for i in 1..HALF_N {
+            right[i] -= even_s_conv[i] + left[i];
+            even_s_conv[i] += left[i - 1];
+        }
+
+        // Interleave even_s_conv and right in the output:
+        for i in 0..HALF_N {
+            output[2 * i] = even_s_conv[i];
+            output[2 * i + 1] = output[i + HALF_N];
+        }
+    }
+
+    #[inline(always)]
+    fn conv3(lhs: [T; 3], rhs: [U; 3], output: &mut [V]) {
+        output[0] = Self::parity_dot(lhs, [rhs[0], rhs[2], rhs[1]]);
+        output[1] = Self::parity_dot(lhs, [rhs[1], rhs[0], rhs[2]]);
+        output[2] = Self::parity_dot(lhs, [rhs[2], rhs[1], rhs[0]]);
+    }
+
+    #[inline(always)]
+    fn signed_conv3(lhs: [T; 3], rhs: [U; 3], output: &mut [V]) {
+        output[0] = Self::parity_dot(lhs, [rhs[0], -rhs[2], -rhs[1]]);
+        output[1] = Self::parity_dot(lhs, [rhs[1], rhs[0], -rhs[2]]);
+        output[2] = Self::parity_dot(lhs, [rhs[2], rhs[1], rhs[0]]);
+    }
+
+    #[inline(always)]
```

Comment:
Probably not worth changing but just wanted to point out that conv4 is exactly implementing:

conv_n::<4, 2, _, _>(lhs, rhs, output, Self::conv2, Self::signed_conv2)